sdk/core/azure-core/src/main/java/com/azure/core/util/Configuration.java

```diff
@@ -375,66 +340,131 @@ public String remove(String name) {
-     * @param defaultValue Default value to return if the configuration value is null or empty.
-     * @param <T> Generic type that the value is converted to if not null or empty.
-     * @return The converted configuration, if null or empty the default value.
-     */
-    @SuppressWarnings("unchecked")
-    private static <T> T convertOrDefault(String value, T defaultValue) {
-        // Value is null or empty, return the default.
-        if (CoreUtils.isNullOrEmpty(value)) {
-            return defaultValue;
+     * <!-- src_embed com.com.azure.core.util.Configuration#get(ConfigurationProperty) -->
+     * <!-- end com.azure.core.util.Configuration#get(ConfigurationProperty) -->
+     *
+     * @param property instance.
+     * @param <T> Type that the configuration is converted to if found.
+     * @return true if property is available, false otherwise.
+     * @throws NullPointerException when property instance is null.
+     * @throws IllegalArgumentException when required property is missing.
+     * @throws RuntimeException when property value conversion (and validation) throws.
+     */
+    public <T> T get(ConfigurationProperty<T> property) {
+        Objects.requireNonNull(property, "'property' cannot be null");
+        String value = getWithFallback(property);
+
+        if (value == null) {
+            if (property.isRequired()) {
+                throw logger.atError()
+                    .addKeyValue("name", property.getName())
+                    .addKeyValue("path", path)
+                    .log(new IllegalArgumentException("Missing required property."));
+            }
+            return property.getDefaultValue();
         }
 
-        // Check the default value's type to determine how it needs to be converted.
-        Object convertedValue;
-        if (defaultValue instanceof Byte) {
-            convertedValue = Byte.parseByte(value);
-        } else if (defaultValue instanceof Short) {
-            convertedValue = Short.parseShort(value);
-        } else if (defaultValue instanceof Integer) {
-            convertedValue = Integer.parseInt(value);
-        } else if (defaultValue instanceof Long) {
-            convertedValue = Long.parseLong(value);
-        } else if (defaultValue instanceof Float) {
-            convertedValue = Float.parseFloat(value);
-        } else if (defaultValue instanceof Double) {
-            convertedValue = Double.parseDouble(value);
-        } else if (defaultValue instanceof Boolean) {
-            convertedValue = Boolean.parseBoolean(value);
-        } else {
-            convertedValue = value;
+        try {
+            return property.getConverter().apply(value);
+        } catch (Throwable t) {
+            throw logger.atError()
+                .addKeyValue("name", property.getName())
+                .addKeyValue("value", property.canLogValue() ? value : "redacted")
+                .log(Exceptions.propagate(t));
         }
```

Comment:
We should not swallow or convert `Error`s. We should either make this catch block more specific or properly handle Errors. See https://github.com/Azure/azure-sdk-for-java/pull/27561 for reference.